src/Dapr.Messaging/PublishSubscribe/PublishSubscribeReceiver.cs

```diff
@@ -0,0 +1,261 @@
+﻿// ------------------------------------------------------------------------
+// Copyright 2024 The Dapr Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//     http://www.apache.org/licenses/LICENSE-2.0
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+// ------------------------------------------------------------------------
+
+using System.Runtime.CompilerServices;
+using System.Threading.Channels;
+using Grpc.Core;
+using C = Dapr.AppCallback.Autogen.Grpc.v1;
+using P = Dapr.Client.Autogen.Grpc.v1;
+
+namespace Dapr.Messaging.PublishSubscribe;
+
+/// <summary>
+/// A thread-safe implementation of a receiver for messages from a specified Dapr publish/subscribe component and
+/// topic.
+/// </summary>
+internal sealed class PublishSubscribeReceiver : IAsyncDisposable
+{
+    /// <summary>
+    /// Maintains the stream connection to the Dapr sidecar for the subscription.
+    /// </summary>
+    private readonly ConnectionManager connectionManager;
+
+    /// <summary>
+    /// The name of the Dapr pubsub component.
+    /// </summary>
+    private readonly string pubsubName;
+    /// <summary>
+    /// The name of the topic to subscribe to.
+    /// </summary>
+    private readonly string topicName;
+    /// <summary>
+    /// Options allowing the behavior of the receiver to be configured.
+    /// </summary>
+    private readonly DaprSubscriptionOptions options;
+    /// <summary>
+    /// A channel used to decouple the messages received from the sidecar to their consumption.
+    /// </summary>
+    private readonly Channel<TopicMessage> channel = Channel.CreateUnbounded<TopicMessage>();
+    /// <summary>
+    /// A collection of <see cref="TaskCompletionSource"/> used to signal acknowledgement of received messages so a status
+    /// can be sent back to the sidecar indicating what behavior should happen to each.
+    /// </summary>
+    private readonly Dictionary<string, TaskCompletionSource<bool>> acknowledgementTasks = new();
```

Comment:
I wonder if a `ConcurrentDictionary` could avoid the acknowledgement locks.

Reply:
Solid idea. Replaced here and in the `WaitForAcknowledgementAsync` method.